app/src/main/java/com/babylon/wallet/android/domain/usecases/BuildDappResponseUseCase.kt

```diff
@@ -99,55 +109,120 @@ open class BuildDappResponseUseCase(private val rolaClient: ROLAClient) {
-    private val rolaClient: ROLAClient
-) : BuildDappResponseUseCase(rolaClient) {
+    private val rolaClient: ROLAClient,
+    accessFactorSourcesProxy: AccessFactorSourcesProxy
+) : BuildDappResponseUseCase(accessFactorSourcesProxy = accessFactorSourcesProxy) {
 
-    @Suppress("LongParameterList", "ReturnCount")
+    @Suppress("LongParameterList", "ReturnCount", "LongMethod")
     suspend operator fun invoke(
         request: AuthorizedRequest,
         selectedPersona: Persona,
         oneTimeAccounts: List<Account>,
         ongoingAccounts: List<Account>,
         ongoingSharedPersonaData: PersonaData? = null,
-        onetimeSharedPersonaData: PersonaData? = null,
-        deviceBiometricAuthenticationProvider: suspend () -> Boolean = { true }
+        onetimeSharedPersonaData: PersonaData? = null
     ): Result<WalletToDappInteractionResponse> {
-        val loginWithChallenge = request.authRequest is AuthorizedRequest.AuthRequest.LoginRequest.WithChallenge
+        var entitiesWithSignatures: Map<ProfileEntity, SignatureWithPublicKey> = emptyMap()
+
+        val dappInteractionPersona = DappWalletInteractionPersona(
+            identityAddress = selectedPersona.address,
+            label = selectedPersona.displayName.value
+        )
+
         val authResponse: Result<WalletToDappInteractionAuthRequestResponseItem> =
-            buildAuthResponseItem(request, selectedPersona, deviceBiometricAuthenticationProvider)
-        if (authResponse.isSuccess) {
-            val authProvider = if (loginWithChallenge) {
-                { true } // TODO this will go away with MFA, don't ask for biometrics twice
-            } else {
-                deviceBiometricAuthenticationProvider
+            when (val authRequest = request.authRequest) {
+                is AuthorizedRequest.AuthRequest.LoginRequest.WithChallenge -> {
+                    var response: Result<WalletToDappInteractionAuthRequestResponseItem> = Result.failure(
+                        RadixWalletException.DappRequestException.FailedToSignAuthChallenge()
+                    )
+
+                    getSignaturesForAllEntities(
+                        challenge = authRequest.challenge,
+                        metadata = request.metadata,
+                        entities = getAccountsWithSignatureRequired(
+                            request = request,
+                            ongoingAccounts = ongoingAccounts,
+                            oneTimeAccounts = oneTimeAccounts
+                        ) + selectedPersona.asProfileEntity()
+                    ).onSuccess { entitiesWithSignaturesResult ->
+                        entitiesWithSignatures = entitiesWithSignaturesResult
+
+                        val signatureForPersona = entitiesWithSignatures[selectedPersona.asProfileEntity()]
+                            ?: return Result.failure(RadixWalletException.DappRequestException.FailedToSignAuthChallenge())
```

Comment:
you can return `Result.failure(response)` using response declared above.

Reply:
not really because this `return` returns a type of `WalletToDappInteractionResponse`